README.md
[![Workflow Status](https://github.com/fosskers/totp/workflows/Go/badge.svg)](https://github.com/fosskers/totp/actions?query=workflow%3A%22Go%22)
[![Go Report Card](https://goreportcard.com/badge/github.com/fosskers/totp)](https://goreportcard.com/report/github.com/fosskers/totp)
[![PkgGoDev](https://pkg.go.dev/badge/github.com/fosskers/totp)](https://pkg.go.dev/github.com/fosskers/totp)

# totp

A simple, correct TOTP library.

Time-based One-time Passwords are a useful way to authenticate a client,
since a valid password expires long before it could ever be guessed by an
attacker. This library provides an implementation of TOTP that matches its
specification [RFC6238], along with a simple interface.

## Usage

The `Totp` function is likely what you need. It uses the default time step
of 30 seconds and produces 8 digits of output:

```go
 // Negotiated between you and the authenticating service.
 password := []byte("secret")

 // The number of seconds since the Unix Epoch.
 seconds := uint64(time.Now().Unix())

 // Specify the desired Hash algorithm from the Standard Library.
 // For TOTP, sha1 and sha256 are also valid.
 totp := Totp(sha512.New, password, seconds)
```

For full control over how the algorithm is configured, consider
`TotpCustom`.

## Resources
- [RFC6238: TOTP][RFC6238]
- [RFC6238 Errata](https://www.rfc-editor.org/errata_search.php?rfc=6238)

[RFC6238]: https://tools.ietf.org/html/rfc6238

License: MIT


CHANGELOG.md
# totp Changelog

## 1.0.1 (2020-08-22)

#### Fixed

- A bug that prevented the `digits` argument from being respected.

## 1.0.0 (2020-08-21)

Initial release.


totp_test.go
package totp

import (
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"fmt"
	"hash"
	"testing"
	"time"
)

// --- Unit Tests --- //

func testBytes(t *testing.T, expected [8]byte, result [8]byte) {
	if result != expected {
		t.Errorf("toBytes: expected ?, got ?") // TODO How to print byte arrays?
	}
}

func TestToBytes(t *testing.T) {
	testBytes(t, [8]byte{0, 0, 0, 0, 0, 0, 0, 1}, toBytes(59/DEFAULT_STEP))
	testBytes(t, [8]byte{0, 0, 0, 0, 0x02, 0x35, 0x23, 0xec}, toBytes(1111111109/DEFAULT_STEP))
	testBytes(t, [8]byte{0, 0, 0, 0, 0x27, 0xbc, 0x86, 0xaa}, toBytes(20000000000/DEFAULT_STEP))
}

func testTotp(t *testing.T, h func() hash.Hash, secret []byte, time uint64, expected string) {
	hash := Totp(h, secret, time)
	if hash != expected {
		t.Errorf("Totp: expected %s, got %s", expected, hash)
	}
}

func TestTotp1(t *testing.T) {
	secret := []byte("12345678901234567890")
	if len(secret) != 20 {
		t.Errorf("Byte Conversion: expected 20, got %d", len(secret))
	}

	testTotp(t, sha1.New, secret, 59, "94287082")
	testTotp(t, sha1.New, secret, 1111111109, "07081804")
	testTotp(t, sha1.New, secret, 1111111111, "14050471")
	testTotp(t, sha1.New, secret, 1234567890, "89005924")
	testTotp(t, sha1.New, secret, 2000000000, "69279037")
	testTotp(t, sha1.New, secret, 20000000000, "65353130")
}

func TestTotp256(t *testing.T) {
	secret := []byte("12345678901234567890123456789012")
	if len(secret) != 32 {
		t.Errorf("Byte Conversion: expected 32, got %d", len(secret))
	}

	testTotp(t, sha256.New, secret, 59, "46119246")
	testTotp(t, sha256.New, secret, 1111111109, "68084774")
	testTotp(t, sha256.New, secret, 1111111111, "67062674")
	testTotp(t, sha256.New, secret, 1234567890, "91819424")
	testTotp(t, sha256.New, secret, 2000000000, "90698825")
	testTotp(t, sha256.New, secret, 20000000000, "77737706")
}

func TestTotp512(t *testing.T) {
	secret := []byte("1234567890123456789012345678901234567890123456789012345678901234")
	if len(secret) != 64 {
		t.Errorf("Byte Conversion: expected 64, got %d", len(secret))
	}

	testTotp(t, sha512.New, secret, 59, "90693936")
	testTotp(t, sha512.New, secret, 1111111109, "25091201")
	testTotp(t, sha512.New, secret, 1111111111, "99943326")
	testTotp(t, sha512.New, secret, 1234567890, "93441116")
	testTotp(t, sha512.New, secret, 2000000000, "38618901")
	testTotp(t, sha512.New, secret, 20000000000, "47863826")
}

func TestVariableLength(t *testing.T) {
	secret := []byte("12345678901234567890123456789012")
	hash := TotpCustom(sha256.New, DEFAULT_STEP, 10, secret, 100)
	if hash != "2102975832" {
		t.Errorf("Expected 10 digits, got the hash: %s", hash)
	}

	hash = TotpCustom(sha256.New, DEFAULT_STEP, 6, secret, 100)
	if hash != "975832" {
		t.Errorf("Expected 6 digits, got the hash: %s", hash)
	}
}

// --- Documentation Examples --- //

func ExampleTotp() {
	password := []byte("secret")
	seconds := uint64(time.Now().Unix())
	totp := Totp(sha512.New, password, seconds)
	fmt.Println(totp)
}

// This example uses a step size of 15 seconds (i.e. the password would expire
// faster) and a digit count of 6. Consider DEFAULT_STEP and DEFAULT_DIGITS if
// you only need to alter one of the arguments.
func ExampleTotpCustom() {
	password := []byte("secret")
	seconds := uint64(time.Now().Unix())
	totp := TotpCustom(sha512.New, 15, 6, password, seconds)
	fmt.Println(totp)
}


totp.go
// A simple, correct TOTP implementation.
//
// Time-based One-time Passwords are a useful way to authenticate a client,
// since a valid password expires long before it could ever be guessed by an
// attacker. This library provides an implementation of TOTP that matches its
// specification (RFC6238), along with a simple interface.
//
// Usage
//
// The Totp function is likely what you need. It uses the default time step of
// 30 seconds and gives 8 digits of output:
//
//   // Negotiated between you and the authenticating service.
//   password := []byte("secret")
//
//   // The number of seconds since the Unix Epoch.
//   seconds := uint64(time.Now().Unix())
//
//   // Specify the desired Hash algorithm from the Standard Library.
//   // For TOTP, sha1 and sha256 are also valid.
//   totp := Totp(sha512.New, password, seconds)
//
// Resources
//
// See RFC6239 (https://tools.ietf.org/html/rfc6238) and its Errata
// (https://www.rfc-editor.org/errata_search.php?rfc=6238) for more information.
package totp

import (
	"crypto/hmac"
	"fmt"
	"hash"
)

// DEFAULT_STEP represents a 30 second time step.
const DEFAULT_STEP uint64 = 30

// DEFAULT_DIGITS represents 8 digits of final output.
const DEFAULT_DIGITS uint32 = 8

// Totp produces a Time-based One-time Password with default settings.
func Totp(h func() hash.Hash, password []byte, time uint64) string {
	return TotpCustom(h, DEFAULT_STEP, DEFAULT_DIGITS, password, time)
}

// TotpCustom produces a Time-based One-time Password with full control over
// algorithm parameters.
func TotpCustom(h func() hash.Hash, step uint64, digits uint32, password []byte, time uint64) string {
	// Hash the secret and the time together.
	mac := hmac.New(h, password)
	bytes := toBytes(time / step)
	mac.Write(bytes[:])
	hash := mac.Sum(nil)

	// Magic from the RFC that produces our final value.
	offset := hash[len(hash)-1] & 0xf
	binary := (uint64(hash[offset]&0x7f) << 24) | (uint64(hash[offset+1]) << 16) | (uint64(hash[offset+2]) << 8) | uint64(hash[offset+3])

	return fmt.Sprintf("%0[2]*[1]d", binary%intPow10(uint64(digits)), digits)
}

// Convert an unsigned 64-bit int into its 8 individual bytes.
func toBytes(n uint64) [8]byte {
	const mask uint64 = 0x00000000000000ff
	bytes := [8]byte{0, 0, 0, 0, 0, 0, 0, 0}

	for i := range bytes {
		bytes[7-i] = byte(mask & (n >> (i * 8)))
	}

	return bytes
}

// intPow10 is similar to `Pow`, but operates strictly with integers.
func intPow10(power uint64) uint64 {
	if power == 0 {
		return 1
	}

	var result uint64 = 10
	for i := power; i > 1; i-- {
		result *= 10
	}
	return result
}
